File: src/utils/position.cpp
#include "utils/position.hpp"

#include <cmath>
#include <ostream>
#include <string>
#include <utility>

namespace dm {

// ================================================================================================
// | CONSTRUCTORS & DESTRUCTORS |
// ==============================

    Position::Position(void)
        : Position(0, 0)
    {}
            
    Position::Position(
        std::size_t row,
        std::size_t column
    ) {
        this->setRow(row);
        this->setColumn(column);
    }
    
// ================================================================================================
// | ACCESSORS |
// =============

    std::size_t Position::getRow(void) const {
        return this->_row;
    }
    
    std::size_t Position::getColumn(void) const {
        return this->_column;
    }

// ================================================================================================
// | MODIFIERS |
// =============

    void Position::setRow(std::size_t row) {
        this->_row = row;
    }
    
    void Position::setColumn(std::size_t column) {
        this->_column = column;
    }
    
// ================================================================================================
// | COMPARATORS |
// ===============

    bool Position::equals(const Position& other) const {
        return this->compare(other) == 0;
    }
    
    std::size_t Position::compare(const Position& other) const {        
        // order by row then column
        if (this->getRow() != other.getRow()) {
            return this->getRow() - other.getRow();
        }
        return this->getColumn() - other.getColumn();;
    }

// ========================================================================================
// | CONVERTERS |
// ==============

    std::string Position::toString(void) const {
        return std::string("(")
            + std::to_string(this->getRow()) + ", "
            + std::to_string(this->getColumn()) + ")";
    }

// ================================================================================================
// | OPERATORS |
// =============
    
    bool operator==(
        const Position& positionA,
        const Position& positionB
    ) {
        return positionA.equals(positionB);
    }
    
    bool operator<(
        const Position& positionA,
        const Position& positionB
    ) { 
        return positionA.compare(positionB) < 0;
    }
    
    bool operator<=(
        const Position& positionA,
        const Position& positionB
    ) {
        return positionA.compare(positionB) <= 0;
    }
    
    bool operator>(
        const Position& positionA,
        const Position& positionB
    ) {
        return positionA.compare(positionB) > 0;
    }

    bool operator>=(
        const Position& positionA,
        const Position& positionB
    ) {
        return positionA.compare(positionB) >= 0;
    }
    
    std::ostream& operator<<(
        std::ostream& stream, 
        const Position& position
    ) {
        stream << position.toString();
        return stream;
    }
}

File: src/dm/base/Position.java
package dm.base;

public final class Position implements Cloneable, Comparable<Position> {

// ================================================================================================
// | FIELDS |
// ==========

    private final int row;
    private final int column;
    private final int layer;

// ================================================================================================
// | CONSTRUCTORS |
// ================

    public Position() { this(0, 0); }
    public Position(int row, int column) { this(row, column, 0) }
    public Position(int row, int column, int layer) {
        this.row = row;
        this.column = column;
        this.layer = layer;
    }
    public Position(Position position) {
        this(
            position.getRow(),
            position.getColumn(),
            position.getLayer()
        );
    }

// ================================================================================================
// | ACCESSORS |
// =============

    public int getRow() { return this.row; }
    public int getColumn() { return this.column; }
    public int getLayer() { return this.layer; }
    
// ================================================================================================
// | VERIFIERS |
// =============

    public boolean isBoundedBy(Bounds bounds) {
        if (bounds == null) { return false; }
        return bounds.encloses(this);
    }

// ================================================================================================
// | COMPARATORS |
// ===============

    @Override
    public boolean equals(Position position) {
        if (position == null) { return false; }
        return (
            this.getRow() == position.getRow() &&
            this.getColumn() == position.getColumn() &&
            this.getLayer() == position.getLayer()
        );
    }
    
    @Override
    public boolean compareTo(Position position) {
        if (this.getRow() != position.getRow()) {
            return this.getRow() - position.getRow();
        } else if (this.getColumn() != position.getColumn()) {
            return this.getColumn() - position.getColumn();
        } else {
            return this.getLayer() - position.getLayer();
        }
    }


// ================================================================================================
// | REPLICATORS |
// ===============

    @Override
    public Bounds clone() { return new Position(this); }

// ================================================================================================
// | CONVERTERS |
// ==============

    @Override
    public String toString() {
        return "("
            + this.getRow() + ", "
            + this.getColumn() + ", "
            + this.getLayer() + ")";
    }
}

